### PR TITLE
Lecture4

### PR DESCRIPTION
**Pull request by carmenhorendum**

### CODE REVIEW

File: Lecture4/src/main/java/App.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+		while (!newOperator.equals(Operator.QUIT)) {
+			System.out.println("Enter simple expression in 'number-operator-number' format, e.g. '2 plus 2'.");
+			System.out.println("=======Operator=======");
+			System.out.println("> Type 'plus' to Add");
+			System.out.println("> Type 'minus' to Subtract");
+			System.out.println("> Type 'multiply' to Multiply");
+			System.out.println("> Type 'divide' to Divide");
+			System.out.println("> Type 'quit' to Exit");
+			System.out.println("======================");
+			Validator validator = new Validator();
```
Comment (nemec8g) [inline Lecture4/src/main/java/App.java]: `Validator validator = new Validator();` - to static final field 
you shouldn't create  new instance each iteration

File: Lecture4/src/main/java/App.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+public class App {
+	public static void main(String[] args) {
+		Operator newOperator = Operator.RUN;
```
Comment (nemec8g) [inline Lecture4/src/main/java/App.java]: newOperator -> operator

File: Lecture4/src/main/java/Calculator.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			case MINUS:
+				total = operand1 - operand2;
+				break;
+			case MULTIPLY:
+				total = operand1 * operand2;
+				break;
+			case DIVIDE:
+				total = division();
+				break;
+//			case QUIT:
```
Comment (nemec8g) [inline Lecture4/src/main/java/Calculator.java]: re move commented code

File: Lecture4/src/main/java/Validator.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import java.util.InputMismatchException;
+import java.util.Scanner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Validator {
```
Comment (nemec8g) [inline Lecture4/src/main/java/Validator.java:5]: it is not validator
this class has two responsibility : 
- read value from console
 - check correctness 

i suggest to move `validate` operation to separate validator (just validate responsibility) 
and rename this class to Reader

File: Lecture4/src/main/java/Validator.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+	public double getOperand() {
+		return operand;
+	}
+
+	public Operator getOperator() {
+		return operator;
+	}
+
+	public boolean isNumber() {
```
Comment (nemec8g) [inline Lecture4/src/main/java/Validator.java]: you don't use result of your methods
i suggest to rename it to ` int getNumbers()`  and return result to main class - and use it there

File: Lecture4/src/test/java/CalcUnitTests.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	@BeforeClass
+	public void oneTimeSetUp() {
+		newCalc();
+		System.out.print("Before Class: newCalc();");
+	}
+
+	@Test(dataProvider = "testAddDataProvider")
+	public void add_twoNumbers_returnsSum(double operand1, double operand2, double total) {
```
Comment (nemec8g) [inline Lecture4/src/test/java/CalcUnitTests.java]: not java convention naming 
`addTwoNumbersReturnsSum ` or better `checkSum`

File: Lecture4/src/test/java/CalcUnitTests.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+		operator = Operator.PLUS;
+		op1 = getOperand(operand1);
```
Comment (nemec8g) [inline Lecture4/src/test/java/CalcUnitTests.java]: what for you use private method which returns single its own param? 
just Calculator calc=  new Calculator(operand1 Operator.PLUS, operand2);
Assert.assertEquals(calc.makeCalculation(), total);

please fix it in all tests

File: Lecture4/src/test/java/CalcUnitTests.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class CalcUnitTests {
+	private double op1;
```
Comment (nemec8g) [inline Lecture4/src/test/java/CalcUnitTests.java]: all tests should be independent
so do not use this private fields
you already have all what you need it test parameters

File: Lecture4/src/test/java/CalcUnitTests.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+	@Test(expectedExceptions = IllegalArgumentException.class,
+			expectedExceptionsMessageRegExp = "Error: Can't divide by zero!",
+			dataProvider = "testDivByZeroDataProvider")
+	public void div_byZero_returnsIllegalArgumentException(double operand1, double operand2, double total)
+			throws IllegalArgumentException {
+		operator = Operator.DIVIDE;
+		op1 = getOperand(operand1);
+		op2 = getOperand(operand2);
+		System.out.println("New IllegalArgumentException was thrown");
```
Comment (nemec8g) [inline Lecture4/src/test/java/CalcUnitTests.java]: IllegalArgumentException will be throw only on next line. so this line is not correct - just remove

File: Lecture4/src/main/java/App.java (hunk 1 above)
Comment (carmenhorendum) [reply Lecture4/src/main/java/App.java]: fixed

File: Lecture4/src/main/java/App.java (hunk 2 above)
Comment (carmenhorendum) [reply Lecture4/src/main/java/App.java]: fixed

File: Lecture4/src/main/java/Calculator.java (hunk 3 above)
Comment (carmenhorendum) [reply Lecture4/src/main/java/Calculator.java]: done

File: Lecture4/src/main/java/Validator.java (hunk 4 above)
Comment (carmenhorendum) [reply Lecture4/src/main/java/Validator.java:5]: done

File: Lecture4/src/main/java/Validator.java (hunk 5 above)
Comment (carmenhorendum) [reply Lecture4/src/main/java/Validator.java]: done

File: Lecture4/src/test/java/CalcUnitTests.java (hunk 6 above)
Comment (carmenhorendum) [reply Lecture4/src/test/java/CalcUnitTests.java]: fixed

File: Lecture4/src/test/java/CalcUnitTests.java (hunk 7 above)
Comment (carmenhorendum) [reply Lecture4/src/test/java/CalcUnitTests.java]: done

File: Lecture4/src/test/java/CalcUnitTests.java (hunk 8 above)
Comment (carmenhorendum) [reply Lecture4/src/test/java/CalcUnitTests.java]: fixed

File: Lecture4/src/test/java/CalcUnitTests.java (hunk 9 above)
Comment (carmenhorendum) [reply Lecture4/src/test/java/CalcUnitTests.java]: fixed

File: Lecture4/src/test/java/CalcUnitTests.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class CalcUnitTests {
+
+	@BeforeClass
+	public void oneTimeSetUp() {
+		System.out.print("Before Class");
```
Comment (nemec8g) [inline Lecture4/src/test/java/CalcUnitTests.java]: no need to create empty baseClass 
you may create new Calculator and check that it is not null

File: Lecture4/src/test/java/CalcUnitTests.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+        @Test(dataProvider = "testAddDataProvider")
+        public void checkSum(double operand1, double operand2, double total) {
+            Calculator calculator = new Calculator(operand1, Operator.PLUS, operand2);
+            System.out.println("Test Total: " + calculator.makeCalculation());
```
Comment (nemec8g) [inline Lecture4/src/test/java/CalcUnitTests.java]: you did action twice 
calculator.makeCalculation() - to local var `results` - and use

File: Lecture4/src/test/java/CalcUnitTests.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+				{2, -2, -4},
+				{2, 22, 44}
+		};
+	}
+
+        @Test(expectedExceptions = IllegalArgumentException.class,
+                expectedExceptionsMessageRegExp = "Error: Can't divide by zero!",
+                dataProvider = "testDivByZeroDataProvider")
+        public void checkDivisionByZero(double operand1, double operand2, double total)
+                throws IllegalArgumentException {
```
Comment (nemec8g) [inline Lecture4/src/test/java/CalcUnitTests.java]: this method should not throw `IllegalArgumentException`
testNg will catch it

File: Lecture4/src/test/java/CalcUnitTests.java (hunk 10 above)
Comment (carmenhorendum) [reply Lecture4/src/test/java/CalcUnitTests.java]: removed

File: Lecture4/src/test/java/CalcUnitTests.java (hunk 11 above)
Comment (carmenhorendum) [reply Lecture4/src/test/java/CalcUnitTests.java]: done

File: Lecture4/src/test/java/CalcUnitTests.java (hunk 12 above)
Comment (carmenhorendum) [reply Lecture4/src/test/java/CalcUnitTests.java]: fixed

**Pull request merged**